daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Struct.scala

```diff
@@ -8,33 +8,48 @@ import IdString.`Name order instance`
 import Ref.Name
 import ScalazEqual._
 
-import scala.collection
 import scalaz.std.iterable._
 import scalaz.std.tuple._
 import scalaz.syntax.order._
 import scalaz.{Equal, Order}
 
-/** We use this container to describe structural record as sorted flat list in various parts of the codebase.
-  *    `entries` are sorted by their first component without duplicate.
-  */
-final case class Struct[+X] private (private val sortedFields: ImmArray[(Ref.Name, X)])
-    extends NoCopy {
+import scala.annotation.tailrec
+import scala.collection.immutable._
 
-  /** O(n) */
+final class Struct[+X] private (protected val sortedFields: ArraySeq[(Ref.Name, X)])
+    extends AnyVal {
+
+  /** O(log n) */
   @throws[IndexOutOfBoundsException]
   def apply(name: Ref.Name): X = sortedFields(indexOf(name))._2
 
-  /** O(n) */
-  def indexOf(name: Ref.Name): Int = sortedFields.indexWhere(_._1 == name)
-
-  /** O(n) */
-  def lookup(name: Ref.Name): Option[X] = sortedFields.find(_._1 == name).map(_._2)
+  /** O(log n) */
+  def indexOf(name: Ref.Name): Int = {
```

Comment:
We probably should extend our unit tests. We only test up to size 3 which for a binary search is very little.

Reply:
b97c33940765ad67c6c32dc11a39d90c25796542